### PR TITLE
EnC syntax matching for local functions

### PR DESCRIPTION
**Pull request by ivanbasov**
Towards https://github.com/dotnet/roslyn/issues/5156

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs [hunk 1]
```diff
@@ -863,6 +863,323 @@ public void LocalFunctionDefinitions()
...
+            {
+                { "a += async u => u;", "a += localF;" },
+                { "async u => u", "object localF(object u) => u;" }
+            };
+
+            expected.AssertEqual(actual);
+        }
+
+        [Fact]
+        public void LocalFunctions4()
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: >public void LocalFunctions4() [](start = 8, length = 29)

I don't think this one is necessary. The original test validates matching of query clauses. There is nothing specific to local functions.

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs [hunk 2]
```diff
@@ -863,6 +863,323 @@ public void LocalFunctionDefinitions()
...
+    return b();
+}
+";
+
+            var matches = GetMethodMatches(src1, src2);
+            var actual = ToMatchingPairs(matches);
+
+            var expected = new MatchingPairs
+            {
+                { "int a()  {     int b()     {         int c()         {             int d() { return 0; };         };          return c();     };      return b(); }", "int a()  {     int b()     {         int c()         {             int d() { return 0; };         };          return c();     };      return b(); }" },
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: >, [](start = 167, length = 1)

I'd insert new lines after coma here and other long lines to make this more readable

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs [hunk 3]
```diff
@@ -863,6 +863,323 @@ public void LocalFunctionDefinitions()
...
+                { "return 0;", "return 0;" },
+                { "return c();", "return c();" },
+                { "return b();", "return b();" }
+            };
+
+            expected.AssertEqual(actual);
+        }
+
+        [Fact]
+        public void LocalFunctions6()
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: >LocalFunctions6 [](start = 20, length = 15)

Do we need this one? What is it testing beyond the previous test?

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs [hunk 4]
```diff
@@ -863,6 +863,323 @@ public void LocalFunctionDefinitions()
...
+    F(G4);
+};
+";
+
+            var matches = GetMethodMatches(src1, src2);
+            var actual = ToMatchingPairs(matches);
+
+            var expected = new MatchingPairs
+            {
+                { "void G6(int a) {      int G5(int c) => /*1*/d;     F(G5);      void G4()     {         void G1(int x) => x;         int G3(int w)         {              int G2(int c) => /*2*/d;             return G2(w);         }         F(G3);         F(G1);     };     F(G4); }", "void G6(int a) {      int G5(int c) => /*1*/d + 1;     F(G5);      void G4()     {         int G3(int w)         {              int G2(int c) => /*2*/d + 1;             return G2(w);         }         F(G3);         F(G1);         int G6(int p) => p *2;         F(G6);     };     F(G4); }" },
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: >, [](start = 283, length = 1)

new line

File: src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs [hunk 5]
```diff
@@ -691,60 +683,68 @@ protected override bool TryComputeWeightedDistance(SyntaxNode leftNode, SyntaxNo
...
-        {
-            GetLambdaParts(leftNode, out var leftParameters, out var leftAsync, out var leftBody);
-            GetLambdaParts(rightNode, out var rightParameters, out var rightAsync, out var rightBody);
+            GetNestedFunctionsParts(leftNode, out var leftParameters, out var leftAsync, out var leftBody, out var leftModifiers, out var leftReturnType, out var leftIdentifier, out var leftTypeParameters);
+            GetNestedFunctionsParts(rightNode, out var rightParameters, out var rightAsync, out var rightBody, out var rightModifiers, out var rightReturnType, out var rightIdentifier, out var rightTypeParameters);
 
             if ((leftAsync.Kind() == SyntaxKind.AsyncKeyword) != (rightAsync.Kind() == SyntaxKind.AsyncKeyword))
             {
                 return 1.0;
             }
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs]: We might want to reconsider this check in future
Could you add a comment with a ling to [https://github.com/dotnet/roslyn/issues/11502](https://github.com/dotnet/roslyn/issues/11502)?

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs [hunk 6]
```diff
@@ -83,7 +83,7 @@ public static void AssertIsBody(SyntaxNode syntax, bool allowLambda)
             if (LambdaUtilities.IsLambdaBody(syntax))
             {
                 Debug.Assert(allowLambda);
-                Debug.Assert(syntax is ExpressionSyntax || syntax is BlockSyntax);
+                Debug.Assert(syntax is ExpressionSyntax || syntax is BlockSyntax || syntax is ArrowExpressionClauseSyntax);
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs]: > || syntax is ArrowExpressionClauseSyntax [](start = 80, length = 41)

I would rather change IsLambdaBody instead. It'd be more consistent with lambdas if we considered the actual expression the body of a local function rather then the arrow expression.

**Comment by tmat** [comment]
:clock1:

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs [hunk 7]
```diff
@@ -863,6 +863,323 @@ public void LocalFunctionDefinitions()
...
+    return F(b());
+}
+";
+
+            var matches = GetMethodMatches(src1, src2);
+            var actual = ToMatchingPairs(matches);
+
+            var expected = new MatchingPairs
+            {
+                { "int a()  {     int b()     {         int c()         {             int d() { return 0; };         };          return c();     };      return b(); }", "int a()  {     int b()     {         int c()         {             int d() { return H(0); };         };          return G(c());     };      return F(b()); }" },
```
Comment (agocke) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: This is also really hard to read.

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs (hunk 6 above)
Comment (agocke) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs]: I think that would be contrary to a lot of the assumptions in other code. In the same way the block is the body of a regular method, not the statement list, the `ArrowExpression` is the body of an expression-bodied local function.

Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs]: Not for EnC analysis. EnC has very specific concept of "body".  

---
In reply to: [132526663](https://github.com/dotnet/roslyn/pull/21396#discussion_r132526663) [](ancestors = 132526663)

Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs]: As you can actually see below in the ```//expression body``` check.

---
In reply to: [132549667](https://github.com/dotnet/roslyn/pull/21396#discussion_r132549667) [](ancestors = 132549667,132526663)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs (hunk 1 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: OK. Thanks!

---
In reply to: [132338076](https://github.com/dotnet/roslyn/pull/21396#discussion_r132338076) [](ancestors = 132338076)

File: src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs (hunk 5 above)
Comment (ivanbasov) [inline src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs]: Thank you!

---
In reply to: [132496834](https://github.com/dotnet/roslyn/pull/21396#discussion_r132496834) [](ancestors = 132496834)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs (hunk 3 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: OK. Let us remove it.

---
In reply to: [132338373](https://github.com/dotnet/roslyn/pull/21396#discussion_r132338373) [](ancestors = 132338373)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs (hunk 2 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: OK. Thanks! Also suggest to remove new lines in the code to improve readability.

---
In reply to: [132338242](https://github.com/dotnet/roslyn/pull/21396#discussion_r132338242) [](ancestors = 132338242)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs (hunk 4 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: ok

---
In reply to: [132495769](https://github.com/dotnet/roslyn/pull/21396#discussion_r132495769) [](ancestors = 132495769)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs (hunk 7 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementMatchingTests.cs]: Thanks! Let me try to remove extra lines from code snippets and add splits in the verifications.

---
In reply to: [132525353](https://github.com/dotnet/roslyn/pull/21396#discussion_r132525353) [](ancestors = 132525353)

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs (hunk 6 above)
Comment (ivanbasov) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs]: When I made changes for extracting bodies from local functions, I understood the meaning of the comment. :-) Thank you!

---
In reply to: [132499185](https://github.com/dotnet/roslyn/pull/21396#discussion_r132499185) [](ancestors = 132499185)

**Pull request closed without merging**